### PR TITLE
feat: machine provision watcher mapper

### PR DESCRIPTION
**Pull request by tlm**
This PR introduces a new generic mapper for watchers that care about machine provisioned storage entities. Machine provision watchers care about watching for changes that involve their net node or machine uuid but the caller that is using the watcher wants to know the id value of the storage entity that has changed.

We know that when the watcher fires a storage entity has changed. This mapper just does the leg work to figure out what has changed and then report back a new set of changes for the storage entity ids which have changed.

#20034 complements this PR with the triggers that will be used into this mapper.

**Notes:**
- Still have to write unit tests.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [ ] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [x] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

Unit tests verify the contract on offer. Until this is used in a watcher no end to end user testing can be done.

## Documentation changes

N/A

## Links

**Jira card:** JUJU-8139

### CODE REVIEW

File: domain/storageprovisioning/service/machinemapper.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// A concrete example would be a machine provisioner concern where the concern
+// is the net node of the machine being watched. Entities are all the storage
+// entities that would be provisioned from this net node context.
+//
+// When an entities life has change compared with the last seen value or an
+// entity has been added or removed will result in the id being returned in the
+// change set.
+//
+// [EntityLifeGetter] provides the latest life values for the concerns entities.
+func EntityLifeWatcherMapperFuc(
```
Comment (hpidcock) [inline domain/storageprovisioning/service/machinemapper.go]:
```suggestion
func EntityLifeWatcherMapperFunc(
```

File: domain/storageprovisioning/service/machinemapper.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+) (eventsource.Mapper, error) {
+	knownLife, err := lifeGetter(ctx)
+	if err != nil {
+		return nil, errors.Errorf(
+			"getting initial storage entity known life values: %w", err,
+		)
+	}
+
+	return func(
+		ctx context.Context, _ []changestream.ChangeEvent,
```
Comment (hpidcock) [inline domain/storageprovisioning/service/machinemapper.go]: We still need to check the change events right? To ensure at least one is for the machine we are watching, even if we use the `lifeGetter` to determine the related entity that changed.

File: domain/storageprovisioning/service/machinemapper.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+// [EntityLifeGetter] provides the latest life values for the concerns entities.
+func EntityLifeWatcherMapperFuc(
+	ctx context.Context, lifeGetter EntityLifeGetter,
+) (eventsource.Mapper, error) {
+	knownLife, err := lifeGetter(ctx)
+	if err != nil {
+		return nil, errors.Errorf(
+			"getting initial storage entity known life values: %w", err,
+		)
+	}
```
Comment (hpidcock) [inline domain/storageprovisioning/service/machinemapper.go]: We have a race condition here, if the key identifying the entity is a natural key, there is a non-zero possibility that we miss recreated entities that share the same natural key.

1. Entity **𝑥** exists.
2. `EntityLifeWatcherMapperFunc` sees **𝑥** as alive/dying/dead.
3. Entity 𝑥 is removed.
4. Initial query for watcher does not see **𝑥**.
5. Entity **𝑥** is recreated and transitions to the same life at step 2.
6. Mapper closure from `EntityLifeWatcherMapperFunc` sees **𝑥** change event, but filters it out.
7. No event is ever fired for **𝑥**.

We can fix this one of three ways:
1. Just comment this function that it must be used with UUIDs only.
2. `EntityLifeWatcherMapperFunc` takes a shared map that the initial query must populate.
3. ~~Don't initialise the cache and have double initial events.~~

EDIT: Option 3 suffers from missed removes.

File: domain/storageprovisioning/service/machinemapper.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+// entities that would be provisioned from this net node context.
+//
+// When an entities life has change compared with the last seen value or an
+// entity has been added or removed will result in the id being returned in the
+// change set.
+//
+// [EntityLifeGetter] provides the latest life values for the concerns entities.
+func EntityLifeWatcherMapperFuc(
+	ctx context.Context, lifeGetter EntityLifeGetter,
+) (eventsource.Mapper, error) {
```
Comment (hpidcock) [inline domain/storageprovisioning/service/machinemapper.go]: Also I think this might need a test.

File: domain/storageprovisioning/service/machinemapper.go (hunk 1 above)
Comment (tlm) [reply domain/storageprovisioning/service/machinemapper.go]: Done & Done

File: domain/storageprovisioning/service/machinemapper.go (hunk 2 above)
Comment (tlm) [reply domain/storageprovisioning/service/machinemapper.go]: that would be the predicate value that is supplied to the watcher. I think this mapper should always assume it is just for the one entity.

File: domain/storageprovisioning/service/machinemapper.go (hunk 4 above)
Comment (tlm) [reply domain/storageprovisioning/service/machinemapper.go]: Have added tests.

File: domain/storageprovisioning/service/machinemapper.go (hunk 3 above)
Comment (tlm) [reply domain/storageprovisioning/service/machinemapper.go]: This is a good find thank you. There is also the case that we need to consider where the initial query identifies an id 𝑥 but between reporting this value and the known set being built x is removed. This would mean no change event is ever raised for x. You could argue that this shouldn't be able to happen as the provisioner controls the life x but there is no way to hold this guarantee.

Based on your findings and mine option 2 is the best place for this. I will write a helper to get the mapper and the initial query namespace query all in one call so they are linked.

Comment (tlm) [reply domain/storageprovisioning/service/machinemapper.go]: I believe this case has been dealt with now. I have introduced `MakeEntityLifePrerequisites` that couples the initial query state to the mapper. This will get rid of our race conditions.

I believe the tests I have introduced for this new type prove that this case is dealt with.

File: domain/storageprovisioning/service/machinemapper.go (hunk 2 above)
Comment (tlm) [reply domain/storageprovisioning/service/machinemapper.go]: Just to add to this. I did consider supporting multiple predicate values for the mapper but the implementation just became overly complex for no reason. We know now that the users of this only ever care about 1 value and so there will be no users of this functionality.

We are also relying on the fact that unit net node values are the same as the machine for IAAS models. If this was to ever diverge then I would reconsider this.

File: domain/storageprovisioning/service/entitylife.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+// the change set.
+//
+// [EntityLifeGetter] provides the latest life values for the concerns entities.
+// This func expect to be provided with a func that provides the initial life
+// values to start performing change detection on top of. This func will be
+// called once on the first invocation of the returned mapper.
+//
+// The returned mapper is not thread safe.
+func EntityLifeMapperFunc(
+	initialLife func(ctx context.Context) (map[string]life.Life, error),
```
Comment (hpidcock) [inline domain/storageprovisioning/service/entitylife.go]:
```suggestion
	initialLife EntityLifeGetter,
```

File: domain/storageprovisioning/service/entitylife.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+		return slices.Collect(maps.Keys(initQueryData)), nil
+	}
+
+	initialLifeForMapper := func(ctx context.Context) (map[string]life.Life, error) {
+		select {
+		case initialData := <-initial:
+			return initialData, nil
```
Comment (hpidcock) [inline domain/storageprovisioning/service/entitylife.go]: Can we add one guard rail please:
```suggestion
		case initialData, ok := <-initial:
			if !ok {
				return nil, errors.ConstError("initial life can only be called once or initial query failed")
			}
			return initialData, nil
```

**Review by hpidcock — APPROVED** [review]

File: domain/storageprovisioning/service/entitylife.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+func EntityLifeMapperFunc(
+	initialLife func(ctx context.Context) (map[string]life.Life, error),
+	lifeGetter EntityLifeGetter,
+) eventsource.Mapper {
+	var haveInitialLife bool
+	var knownLife map[string]life.Life
+	return func(
+		ctx context.Context, _ []changestream.ChangeEvent,
+	) ([]string, error) {
+		if !haveInitialLife {
```
Comment (ycliuhw) [inline domain/storageprovisioning/service/entitylife.go:54]: You can simply check if knownLife == nil directly, no need to define a separate boolean variable.
Alternatively, you could retrieve the initial life earlier, before the return.

File: domain/storageprovisioning/service/entitylife.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+		}
+
+		changes := []string(nil)
+		for k, v := range latestLife {
+			if l, has := knownLife[k]; !has || v != l {
+				changes = append(changes, k)
+			}
+			delete(knownLife, k)
+		}
+		changes = slices.AppendSeq(changes, maps.Keys(knownLife))
```
Comment (ycliuhw) [inline domain/storageprovisioning/service/entitylife.go:79]: `// Append the removed entity changes.`
Better to add a comment.

Comment (ycliuhw) [reply domain/storageprovisioning/service/entitylife.go:79]: Based on this EntityLifeMapperFunc, it seems the watcher implementation will fetch the current life of all entities upon receiving an event, rather than processing the contents of the ChangeEvents. Essentially, the source watcher is treated as a notification trigger.

File: domain/storageprovisioning/service/entitylife_test.go [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+			return nil, errors.Errorf("no more values in sequence")
+		}
+		return vals, nil
+	}, stop
+}
+
+// initialEntityLifeQuery is a helper function that returns an eventsource query
+// that will return the data supplied when called.
+func initialEntityLifeQuery(data map[string]life.Life) eventsource.Query[map[string]life.Life] {
+	return func(_ context.Context, _ database.TxnRunner) (map[string]life.Life, error) {
```
Comment (ycliuhw) [inline domain/storageprovisioning/service/entitylife_test.go]: No need the `_`.

File: domain/storageprovisioning/service/entitylife.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+		ctx context.Context, _ []changestream.ChangeEvent,
+	) ([]string, error) {
+		if !haveInitialLife {
+			l, err := initialLife(ctx)
+			if err != nil {
+				return nil, errors.Errorf("getting initial life for mapper: %w", err)
+			}
+			haveInitialLife = true
+			knownLife = l
+		}
```
Comment (ycliuhw) [inline domain/storageprovisioning/service/entitylife.go:61]: In the first round, we will get the `initialLife` and then immediately call `lifeGetter`, which will return the same result. So we compare the same map value, then return the empty diff. Probably we can do better here.

File: domain/storageprovisioning/service/entitylife_test.go [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+		{
+			Name:        "nil getter values",
+			InitialLife: map[string]life.Life{},
+			LifeStages: []map[string]life.Life{
+				{},
+			},
+			Expected: [][]string{
+				{},
+			},
+		},
```
Comment (ycliuhw) [inline domain/storageprovisioning/service/entitylife_test.go:155]: We need one more test to deal with duplicate events.

```go
{
			Name:        "identical result for duplicated events",
			InitialLife: map[string]life.Life{"fs1": life.Alive},
			LifeStages: []map[string]life.Life{
				{"fs1": life.Alive},
			},
			Expected: [][]string{
				{},
			},
		},
```

**Review by ycliuhw — COMMENTED** [review]
Thanks for the change.
A few questions.

File: domain/storageprovisioning/service/entitylife.go (hunk 10 above)
Comment (ycliuhw) [reply domain/storageprovisioning/service/entitylife.go:61]: Ignore this comment.
We have already ensured the initial data is cached in the channel when the initial life getter gets executed in the NameSpaceWatcher.

File: domain/storageprovisioning/service/entitylife.go (hunk 5 above)
Comment (tlm) [reply domain/storageprovisioning/service/entitylife.go]: Done & Done

File: domain/storageprovisioning/service/entitylife.go (hunk 6 above)
Comment (tlm) [reply domain/storageprovisioning/service/entitylife.go]: Done & Done

File: domain/storageprovisioning/service/entitylife.go (hunk 7 above)
Comment (tlm) [reply domain/storageprovisioning/service/entitylife.go:54]: what happens if the initial query returns nil for the initial life? One problematic implementation could make this fail in subtle ways.

File: domain/storageprovisioning/service/entitylife.go (hunk 8 above)
Comment (tlm) [reply domain/storageprovisioning/service/entitylife.go:79]: yes it is treated as notify watcher with a predicate.

File: domain/storageprovisioning/service/entitylife_test.go (hunk 9 above)
Comment (tlm) [reply domain/storageprovisioning/service/entitylife_test.go]: Done & Done
